make_prg/utils/seq_utils.py

```diff
@@ -0,0 +1,223 @@
+from Bio.Seq import Seq
+
+
+NONMATCH = "*"
+GAP = "-"
+Sequence = str
+Sequences = List[str]
+
+
+def is_non_match(letter: str):
+    return letter == NONMATCH
+
+
+def is_gap(letter: str):
+    return letter == GAP
+
+
+def remove_duplicates(seqs: Sequences) -> Generator:
+    seen = set()
+    for x in seqs:
+        if x in seen:
+            continue
+        seen.add(x)
+        yield x
+
+
+def has_empty_sequence(alignment: MSA, interval: Tuple[int, int]) -> bool:
+    sub_alignment = alignment[:, interval[0] : interval[1] + 1]
+    for record in sub_alignment:
+        if all(map(is_gap, record.seq)):
+            return True
+    return False
+
+
+def ungap(seq: str) -> str:
+    return seq.replace(GAP, "")
+
+
+def count(iterable) -> int:
+    return sum(1 for _ in iterable)
+
+
+def get_alignment_seqs(alignment: MSA) -> Generator:
+    for record in alignment:
+        yield str(record.seq)
+
+
+def get_number_of_unique_ungapped_sequences(sub_alignment: MSA) -> int:
+    alignment_seqs = get_alignment_seqs(sub_alignment)
+    ungapped_sequences = map(ungap, alignment_seqs)
+    deduplicated_ungapped_sequences = remove_duplicates(ungapped_sequences)
+    number_of_unique_nongapped_sequences = count(deduplicated_ungapped_sequences)
+    return number_of_unique_nongapped_sequences
+
+
+def get_number_of_unique_gapped_sequences(sub_alignment: MSA) -> int:
+    alignment_seqs = get_alignment_seqs(sub_alignment)
+    deduplicated_gapped_sequences = remove_duplicates(alignment_seqs)
+    number_of_unique_gapped_sequences = count(deduplicated_gapped_sequences)
+    return number_of_unique_gapped_sequences
+
+
+class SequenceCurationError(Exception):
+    pass
+
+
+class SequenceExpander:
+    iupac = {
+        "R": "GA",
+        "Y": "TC",
+        "K": "GT",
+        "M": "AC",
+        "S": "GC",
+        "W": "AT",
+        "A": "A",
+        "C": "C",
+        "G": "G",
+        "T": "T",
+    }
+    expandable_bases = set(iupac.keys())
+    allowed_bases = expandable_bases.union({"N"})
+    standard_bases = {"A", "C", "G", "T"}
+    ambiguous_bases = expandable_bases.difference(standard_bases)
+
+    @classmethod
+    def check_if_there_is_sequence_with_disallowed_bases(cls, sequences: List[str]):
+        for sequence in sequences:
+            if not set(sequence).issubset(cls.allowed_bases):
+                raise SequenceCurationError(
+                    "A slice of a sequence has a disallowed base.\n"
+                    f"Allowed bases: {cls.allowed_bases}.\n"
+                    f"Sequence: {sequence}\n"
+                    f"Redo sequence curation.\n"
+                )
+
+    @classmethod
+    def check_all_sequences_are_composed_of_ACGT(cls, sequences: List[str]):
+        for sequence in sequences:
+            sequence_is_composed_of_ACGT_only = set(sequence).issubset(cls.standard_bases)
+            assert sequence_is_composed_of_ACGT_only, f"Sequence ({sequence}) should be composed only of ACTG only."
```

Comment:
```suggestion
            assert sequence_is_composed_of_ACGT_only, f"Sequence ({sequence}) should be composed of ACTG only."
```

Reply:
I thought we allowed some small number of ambiguities?

Reply:
We do in the input, but not in the output, i.e. PRG. This is done just after expanding the alignments sequences, and when expanding, we transform non-ACGT bases into ACGT, and thus we should have just ACGT in the output. Also, if we encounter a base that is not ACGT, the binary PRG encoder, that produces input for `gramtools` is going to error out, as it just encodes ACGT (I think also both `gramtools` and `pandora` does not handle ambiguous bases in the PRG...).